Refs [TASK][26285] Local Data Source

toandv98 opened the pull request:
## Related Tickets

- [#TASK 26285](https://edu-redmine.sun-asterisk.vn/issues/26285)

## WHAT

- Databse design, create DatabaseHelper, create Local Data Source

## HOW

- Create DatabaseHelper using SQLiteOpenHelper, create Local Data Source

## WHY

## Evidence (Screenshot or Video)

## Performance (Optional)

## Review Checklist

| Category      | View Point                                                   | Description                                                  | Expected Reviewer Answer | Reviewer1 (ToanDV) | Reviewer2 (HuyTQ)  |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------ | ------------------ | ------------------ |
| General       | Code should be self-explanatory.                             | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Is this function or class too big? If yes, is the function or class having too many responsibilities? |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| General       | Does the code follow DRY?: Do not Repeat Yourself            | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Exception handling and cleanup (dispose) resources           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Does the code build and execute correctly?                   |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the source code have end-line errors?                   |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| Conventions   | No hard coding, use constants/configuration values           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Redmine       | Does the ticket follow Sun* Redmine working process?         | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |

## Notes (Optional)

*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

huytq-2422 commented on app/src/main/java/com/edu/sun/oereminder/data/repository/MessageRepositoryImpl.kt:
```diff
@@ -0,0 +1,134 @@
+    private val appExecutors: AppExecutors
+) : MessageRepository {
+
+    override fun getMessages(callBack: SourceCallBack<List<Message>>) {
```
```suggestion
    override fun getMessages(callback: SourceCallBack<List<Message>>) {
```

huytq-2422 commented on app/build.gradle:
```diff
@@ -35,5 +37,8 @@ dependencies {
     implementation 'com.google.android.material:material:1.3.0-alpha02'
     implementation 'androidx.fragment:fragment:1.3.0-alpha07'
-
+    implementation 'androidx.security:security-crypto:1.1.0-alpha01'
```
Thư viện này dùng để làm gì vậy

huytq-2422 replied:
Dùng trong trường hợp nào của app thế

toandv98 replied:
Thư viện này dùng để mã hóa và quản lý file hặc sharedPreferences.
App của t dùng thư viện này để mã hóa và lưu ChatworkToken vì token của nó là cố định.

huytq-2422 commented on app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt:
```diff
@@ -0,0 +1,36 @@
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+
+const val NUM_FIXED_THREAD = 4
```
hằng số này để làm gì nhỉ?

toandv98 replied:
Đây là kích thước ThreadPool của networkIO Executor. Tôi muốn cùng một lúc chỉ có tối đa 4 thread dùng cho network request. Cũng có thể tính số này theo availableprocessors() nhưng t fixed luôn là 4 cho đơn giản.

toandv98 replied:
T sửa lại lấy theo số processors rồi. không để là 4 nữa

huytq-2422 commented on app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    companion object {
+
+        @JvmStatic
+        fun getInstance() = Holder.INSTANCE
+
+        private object Holder {
+            val INSTANCE = AppExecutors(
+                Executors.newSingleThreadExecutor(),
+                Executors.newFixedThreadPool(NUM_FIXED_THREAD),
+                MainThreadExecutor()
+            )
+        }
+    }
+}
```
```suggestion
    companion object {

        private val lock = Any()
        private var INSTANCE : AppExecutors? = null 

        fun getInstance(): AppExecutors = INSTANCE?: synchronized(lock){
            INSTANCE?: AppExecutors(
                Executors.newSingleThreadExecutor(),
                Executors.newFixedThreadPool(NUM_FIXED_THREAD),
                MainThreadExecutor()
            ).also {
                INSTANCE = it
            }
       }
}
```

toandv98 replied:
Cho t hỏi chút, trường hợp này không truyền parameters vào getInstance thì tại sao không dùng Initialization on Demand mà lại nên dùng Double-checked locking như ô suggest nhỉ.

huytq-2422 commented on app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt:
```diff
@@ -0,0 +1,36 @@
+
+const val NUM_FIXED_THREAD = 4
+
+class AppExecutors private constructor(
+    val diskIO: Executor,
+    val networkIO: Executor,
+    val mainThread: Executor
+) {
```
Class này dùng để làm gì nhỉ, diskIO, networkIO, mainThread là gì

toandv98 replied:
Trong repository tôi dùng diskIO khi truy cập database, networkIO khi request Api, và mainThread để gọi callback trên lên MainThread. trong project này t không dùng AsyncTask.
diskIO và netwokIO là những threadpool khác nhau nên khi threadpool đầy thì truy cập network không phải đợi database và ngược lại. T giải thích hơi bị rối :))

datdt-1935 commented on app/src/main/java/com/edu/sun/oereminder/data/source/local/dbutils/StatementUtils.kt:
```diff
@@ -0,0 +1,45 @@
+    )
+}
+
+data class Where(
```
```suggestion
data class Predicate(
```

toandv98 committed:
Task #26285
 Database design
 Create SQLiteOpenHelper
 Create Local Data Source
 Create repository
